storage: implement simple version of souce-specific statistics

guswynn opened the pull request:
As described here: https://www.notion.so/materialize/User-facing-metrics-for-Sources-9603b61224a04a749969702e04662ed3#ca9365f66fb947128c0608906e556620

The implementation was relatively straightforward, but revealed a few subtleties around how we want to present some data. These are mostly documented in the docs I added to this pr, but I feel we should discuss them here:
- In general I counted retractions and insertions the same.
- `messages_received` and `messages_committed` can diverge because envelopes like `UPSERT` consolidate in within timestamps AND can produce 2 messages to commit (an insert and a retraction).
  - Its possible to move `messages_received` to just track the rows (both retractions and inserts) we are putting in persist batches, but that might is technically a different concept. Do we want two metrics?
- `bytes_received` has numbers that depend on the type of source.
- Its unclear how we want multi-output sources to work. In this pr, I just consolidated them into the same data, but its possible we want different rows for each sub-source.
  - I didn't add a test for multi-output sources because I want to discuss this first
  - Moving to this will likely fix a data-race bug in this pr, where `snapshot_finished` can be updated by the underlying sources too early. 
- Note also that these stats are also reset to 0 on restart...this seems reasonable to me, as this is the same way prometheus works!
- When sources are dropped, I don't yet clear out the data in the shard
  

Overall, the above issues show that we may want to add columns in the future...which means we need to expect users to not materialize `select * from mz_storage_source_statistics`, and that the ui will select specific columns...are these reasonable expectations?

### Motivation


  * This PR adds a known-desirable feature.


### Tips for reviewer
See above.

This pr can be reviewed as separate commits, if you want
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Added `mz_source_statistics`, which contains useful stats about sources

aljoscha commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -1698,6 +1698,22 @@ pub static MZ_STORAGE_HOST_METRICS: Lazy<BuiltinSource> = Lazy::new(|| BuiltinSo
     is_retained_metrics_relation: true,
 });
 
+// When
```
TODO, it seems

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -106,6 +106,10 @@ pub enum IntrospectionType {
     ShardMapping,
     StorageHostMetrics,
+
+    // Note that this single-shard introspection source will be changed to per-replica,
```
nit: I would maybe not say "once storage and compute are merged" but instead sth like "once we allow running/multiplexing multiple sources on a single cluster"

aljoscha commented on src/storage-client/src/client.rs:
```diff
@@ -285,17 +300,35 @@ pub enum StorageResponse<T = mz_repr::Timestamp> {
     /// Punctuation indicates that no more responses will be transmitted for the specified ids
     DroppedIds(Vec<GlobalId>),
+
+    /// A list of statistics updates, currently only for sources
```
This is probably just a pet peeve of mine (and I think Dan's), where we try and use proper capitalization and punctuation in comments.

I know the storage code base is not conforming to this, but the persist code base is quite consistent in this. 😅

guswynn replied:
I am trying to get better at this! usually I copy the style from the other variants, which here is lacking, but will upgrade!

aljoscha commented on src/storage-client/src/client.rs:
```diff
@@ -275,6 +275,21 @@ impl Arbitrary for StorageCommand<mz_repr::Timestamp> {
+/// then control how much `StorageResponse` traffic is produced when sending updates
+/// back to the controller to be written.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct SourceStatisticsUpdate {
```
musing: should we maybe include the (wall-clock, I know ... 🙈) time at which the metrics were produced?

guswynn replied:
I have a comment about this somewhere in the pr...

basically my conclusion is that we should just allow some jitter (note also that the `CollectionManager` we use in the controller to record these doesn't support setting a value at a ts, and adding that would be fairly complex), because this is a simple implementation that is a stopgap; when we move to using logging dataflows and doing this correctly™, we definitely want to be more specific about the times

guswynn replied:
https://github.com/MaterializeInc/materialize/pull/16788/files#r1058531089

aljoscha commented on src/storage-client/src/client.rs:
```diff
@@ -468,6 +520,12 @@ where
                 }
             }
+            StorageResponse::StatisticsUpdates(stats) => {
+                // Just forward it along; the `worker_id` should have been set in `storage_state`.
```
Note to self: talk to Gus about the reasoning for this

guswynn replied:
hello!

aljoscha commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -454,6 +454,37 @@ Field              | Type       | Meaning
+### `mz_storage_source_statistics`
+
+The `mz_storage_source_statistics` table gives the last known storage-specific statistics
+for all processes of all extant storage replicas.
```
I think we said we didn't want to use the term "replica" for these. For now, it seems we can stick to "storage hosts" for this.

aljoscha replied:
Or just use sth like "extant sources and sinks"

guswynn replied:
I copied this from `mz_storage_host_metrics` above haha...will change them both

aljoscha commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -454,6 +454,37 @@ Field              | Type       | Meaning
+- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like postgres,
+  produce deletion messages that we count the same as insertions.
+- `messages_committed` can be permanently lower or higher than `messages_received`, because:
+  - The storage layer can choose to consolidate messages within 1 second
```
nit: one second happens to be the current (default) setting but it might change, so maybe just talk about "within a certain time interval"?

guswynn replied:
will change this!

aljoscha commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -454,6 +454,37 @@ Field              | Type       | Meaning
+- `bytes_recieved` can represent different things for different sources.
+  - For example, it represents the literal size of the key and value from kafka messages, but represents
+  an internal representation of rows for postgres sources.
+- The `messages_` and `bytes_received` metrics can be interpreted as rates that can loosely show
```
typo: `messages_*` ?

aljoscha commented on src/storage-client/src/controller.rs:
```diff
@@ -1462,8 +1481,16 @@ where
-            Some(StorageResponse::StatisticsUpdates(_stats)) => {
-                // TODO: impl
+            Some(StorageResponse::StatisticsUpdates(source_stats)) => {
+                // Note we only hold the lock while moving some plain-old-data around here.
```
Good to include this reasoning! 🙌

aljoscha commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -454,6 +454,37 @@ Field              | Type       | Meaning
+`messages_committed` | [`bigint`]   | How many messages we have committed into our storage layer.
+`bytes_received`     | [`bigint`]   | How many bytes worth of messages we have received from upstream.
+
+These values all require some amount of interpretation:
```
Very thoughtful! \o/

let's talk in person but I think this is close to what we want. Or it!

aljoscha commented on src/storage-client/src/controller/metrics_scraper.rs:
```diff
@@ -7,8 +7,7 @@
 
-//! A tokio task (and support machinery) for maintaining storage-managed
-//! collections.
+//! A tokio task (and support machinery) for producing metrics.
```
Nice catch! But maybe: "for scraping and writing down metrics"?

aljoscha commented on src/storage-client/src/controller/statistics.rs:
```diff
@@ -0,0 +1,106 @@
+use crate::client::SourceStatisticsUpdate;
+use crate::controller::collection_mgmt::CollectionManager;
+
+/// Spawns a task that continually (at an interval) write statistics from storaged's
```
nit: "writes"

aljoscha commented on src/storage-client/src/controller/statistics.rs:
```diff
@@ -0,0 +1,106 @@
+    let (shutdown_tx, mut shutdown_rx) = oneshot::channel::<()>();
+
+    mz_ore::task::spawn(|| "statistics_sender", async move {
+        // Keep track of what we think the contents of the output
```
nit: this sentence seems to be missing some words and/or the syntax seems off.

Also, though: this is from my original code, so my fault! 😅

aljoscha commented on src/storage-client/src/controller/statistics.rs:
```diff
@@ -0,0 +1,106 @@
+        // collection, so that we can emit the required retractions/updates
+        // when we learn about new metrics.
+        //
+        // We assume that one update from the MetricsFetcher contains entries
```
And these descriptions are not correct for the statistics scraper.

aljoscha commented on src/storage-client/src/controller/statistics.rs:
```diff
@@ -0,0 +1,106 @@
+
+/// Spawns a task that continually (at an interval) write statistics from storaged's
+/// that are consolidated in shared memory in the controller.
+pub(super) fn spawn_statistics_sender(
```
maybe `spawn_statistics_scraper` to keep the naming consistent?

aljoscha commented on test/sqllogictest/cluster_log_drop.slt:
```diff
@@ -23,7 +23,7 @@ reset-server
 SELECT (SELECT COUNT(*) FROM mz_catalog.mz_sources WHERE name NOT LIKE '%_1' AND name NOT LIKE '%_2' AND name NOT LIKE '%_3') - (SELECT COUNT(*) FROM mz_catalog.mz_sources WHERE name LIKE '%_1');
 ----
-4
+5
```
yeah, I'm not loving that we have to keep bumping this ... 🙈 

at some point i'll be fed up enough to look into fixing this in a better way... 😅

guswynn replied:
yeah me too lol

aljoscha commented on src/storage/src/render/persist_sink.rs:
```diff
@@ -234,6 +247,11 @@ where
+                            source_statistics.inc_messages_committed_by(
+                                batch_builder.inserts + batch_builder.retractions,
+                            );
+                            source_statistics.update_snapshot_committed(&new_upper);
```
Is this the race you mentioned, and the reason for the boolean in the `Rc`? Because we might otherwise wrongly report `snapshot_committed == false` on a restart?

Not sure I like it, but also don't have a better idea off the top of my head... 😅

guswynn replied:
yeah we need to actually startup the `persist_sink`~

aljoscha reviewed (commented):
Took a thorough look and also left quite some line-level comments. Sorry if the latter are too noisy.

It's nice how (mostly) straightforward this turns out!

aljoscha commented:
@guswynn re "Its unclear how we want multi-output sources to work. In this pr, I just consolidated them into the same data, but its possible we want different rows for each sub-source.":

I think it might make sense to count `messages_received` on the parent source, and let each child source record `messages_committed` (and the other metrics). You see things as they are actually happen, `messages_committed` makes sense to see per sub-source. But you'd have to look at the parent source to see how many messages are coming in in total. WDYT? We can also throw this over to @sjwiesman, @morsapaes, @joacoc, @ggnall to see what they think.

joacoc commented:
> I think it might make sense to count messages_received on the parent source

Agree. Since a source is the initial single starting point, a developer should look for this. The closest to the number of rows in the source, or whatever could be determined as the count, the better. Otherwise, I can imagine myself having to run a query on each sub-source to determine this number.

I'm a bit unsure about what `messages_committed` means, but if it is the number of rows flowing from the parent source to the sub-source looks good.

guswynn commented on src/storage/src/storage_state.rs:
```diff
@@ -230,6 +237,23 @@ impl<'w, A: Allocate> Worker<'w, A> {
 
             self.report_frontier_progress(&response_tx);
 
+            // Note: this interval configures the level of granularity we expect statistics
+            // (at least with this implementation) to have. We expect a statistic in the
+            // system tables to be only accurate to within this interval + whatever
+            // skew the `CollectionManager` adds. The stats task in the controller will
+            // be reporting, for each worker, on some interval,
+            // the statistics reported by the most recent call here. This is known to be
+            // somewhat inaccurate, but people mostly care about either rates, or the
+            // values to within 1 minute.
+            //
+            // TODO(guswynn): Should this be configurable? Maybe via LaunchDarkly?
```
@aljoscha here is where I explain timestamps!

guswynn commented:
@aljoscha and I discussed this, we think that the metrics should be on the sub-sources! The pr is updated to reflect this, and I added a test (load gen sources made this easy)

@aljoscha This pr should be ready to review, I have updated to it to reflect our discussions (and also fixed a bug!!)

@joacoc Yeah, `messages_committed` is the number of messages (row inserts or deletes) ACTUALLY committed to the persist shard for each sub-source

benesch commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -1790,6 +1790,22 @@ pub static MZ_STORAGE_HOST_METRICS: Lazy<BuiltinSource> = Lazy::new(|| BuiltinSo
+    desc: RelationDesc::empty()
+        .with_column("id", ScalarType::String.nullable(false))
+        .with_column("worker_id", ScalarType::UInt64.nullable(false))
+        // These are nullable as we may not fill them all out at the same time.
```
I'm not seeing where this happens—and even if it does, can't we just default them to `false`/`0`?

guswynn replied:
this was from an earlier version, so i can just use `.nullable(true)` here, because they are always set

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+  the changing speed of a source over time. These metrics are periodically reset to 0.
+- In general, the values should be summed across `worker_id`'s, and some metrics may only exist for some workers.
+  - Additionally, sources with multiple outputs (like postgres shards) will need `messages_committed`
+  and `snapshot_committed` summed/AND-ed across the top-level source and the connected tables.
```
Thanks for taking the first cut at this. Looping in @morsapaes to make sure this goes through a docs style pass. The high-level point: this page, more than almost any other page in our docs, is reference documentation. It should deal almost exclusively in hard facts. We don't want to be giving folks advice about how to use the data here—that's better left to guides and tutorials—we should just describe what the data _is_.

guswynn replied:
I definitely think this interpretation section is important, but if it should go somewhere else I'll move it!

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
 `cpu_nano_cores`   | [`bigint`] | Approximate CPU usage, in billionths of a vCPU core.
 `memory_bytes`     | [`bigint`] | Approximate RAM usage, in bytes.
 
+### `mz_storage_source_statistics`
```
"storage source" is not the right name! Either "storage host" or "source". But I think having two separate views—`mz_source_statistics` and `mz_sink_statistics`—is best, since we plan to keep the general structure of these views even after transitioning them to introspection views, and then they'll be consistent with e.g. `mz_{source,sink}_status`.

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+
+Field                 | Type        | Meaning
+----------------------|-------------|--------
+`id`                 | [`text`]     | The ID of the storage object (source or sink).
```
```suggestion
`id`                 | [`text`]     | The ID of the source.
```

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+Field                 | Type        | Meaning
+----------------------|-------------|--------
+`id`                 | [`text`]     | The ID of the storage object (source or sink).
+`worker_id`          | [`bigint`]   | An identifier of a process within a host.
```
```suggestion
`worker_id`          | [`bigint`]   | An identifier of a worker within the source.
```

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
 
+### `mz_storage_source_statistics`
+
+The `mz_storage_source_statistics` table gives the last known storage-specific statistics
+for all processes of all extant storage sources.
```
```suggestion
The `mz_storage_source_statistics` table contains statistics for each process of each source in the system.
```

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+The `mz_storage_source_statistics` table gives the last known storage-specific statistics
+for all processes of all extant storage sources.
+
+At this time, we do not make any guarantees about the exactness or freshness of these numbers.
```
```suggestion
Materialize does not make any guarantees about the exactness or freshness of these numbers.
```

benesch replied:
I think this is the place to mention that these statistics periodically reset to 0 as internal components of the system restart.

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+----------------------|-------------|--------
+`id`                 | [`text`]     | The ID of the storage object (source or sink).
+`worker_id`          | [`bigint`]   | An identifier of a process within a host.
+`snapshot_committed` | [`boolean`]  | Whether or not we have committed the initial snapshot for a source.
+`messages_received`  | [`bigint`]   | How many messages we have received from upstream.
+`messages_committed` | [`bigint`]   | How many messages we have committed into our storage layer.
+`bytes_received`     | [`bigint`]   | How many bytes worth of messages we have received from upstream.
```
It's not clear who "we" is in these descriptions. The worker? The process? The source as a whole? The descriptions of these fields in the design doc used a more precise noun in place of "we" (e.g., "the worker").

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+- `messages_committed` can be permanently lower or higher than `messages_received`, because:
+  - The storage layer can choose to consolidate messages within a certain time interval.
+  - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
+- `bytes_recieved` can represent different things for different sources.
+  - For example, it represents the literal size of the key and value from kafka messages, but represents
+  an internal representation of rows for postgres sources.
```
This feels like too much detail. I would rather add a sentence like "The way in which bytes are counted is specific to each source implementation", or something like that, to the table row, and leave it at that.

benesch commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,39 @@ Field              | Type       | Meaning
+
+These values all require some amount of interpretation:
+
+- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like postgres,
+  produce deletion messages that we count the same as insertions.
+- `messages_committed` can be permanently lower or higher than `messages_received`, because:
+  - The storage layer can choose to consolidate messages within a certain time interval.
+  - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
```
I think we need to go back to the drawing board on this. IMO we can't have `messages_received` and `messages_committed` count messages differently, as they both use "messages" in their name. The metric as described is, I think, better named something like `updates_committed`... but I don't think that's necessarily useful to expose? (In the future, I think `updates_committed` is computable with `SELECT count(*) FROM changes(src)`, and in the meantime I think `SELECT count(*) FROM src` is a decent proxy for `updates_committed`.)

Put another way: the value in exposing these two quantities seemed to be that you could watch for skew between `messages_received and `messages_committed` to detect problems in the source ingestion pipeline. If these two metrics are expected to skew, they seem substantially less useful, and at the very least seems like we need to rename them so that they do not seem so comparable.

aljoscha replied:
strawman: we could have three statistics:
 - `messages_received`: the raw messages that we get from the source
 - `updates` (name tbd) number of updates that we have _after_ messages have gone through the enveloping logic
 - `updates_committed`
 
Depending on the envelope, there will be systemic skew between `messages_received` and `updates`, but with a constant factor/bound. On the other hand,  `updates` and `updates_committed` should proceed in lockstep and you will notice that something is afoot when they don't. wdyt?

aljoscha replied:
cc my usual crew for these user-facing things @morsapaes, @sjwiesman, @joacoc, @ggnall  👋

aljoscha replied:
cc my usual crew for these user-facing things @morsapaes, @sjwiesman, @joacoc, @ggnall  👋

sjwiesman replied:
I think @aljoscha 's take makes sense. Idk what the right names are, I'll need to think about it.

guswynn replied:
@aljoscha `updates` can be incremented when we write the updates into (non-committed) persist batches, so it will always move in lock-step with `updates_committed`

Sounds like a reasonable plan to me!

guswynn replied:
Updated the pr to reflect this (with the given placeholder names)

benesch replied:
I like it! Maybe `updates_staged` instead of `updates`?

morsapaes replied:
+1 for `updates_staged`.

morsapaes commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,37 @@ Field              | Type       | Meaning
+
+Field                 | Type         | Meaning
+----------------------|--------------|--------
+`id`                  | [`text`]     | The ID of the source.
```
```suggestion
`id`                  | [`text`]     | The ID of the source. Corresponds to [mz_catalog.mz_sources.id](https://materialize.com/docs/sql/system-catalog/mz_catalog#mz_sources).
```

morsapaes commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,37 @@ Field              | Type       | Meaning
+Field                 | Type         | Meaning
+----------------------|--------------|--------
+`id`                  | [`text`]     | The ID of the source.
+`worker_id`           | [`bigint`]   | An identifier of a worker within the source.
```
```suggestion
`worker_id`           | [`bigint`]   | The ID of the worker thread.
```

morsapaes commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,37 @@ Field              | Type       | Meaning
+`id`                  | [`text`]     | The ID of the source.
+`worker_id`           | [`bigint`]   | An identifier of a worker within the source.
+`snapshot_committed`  | [`boolean`]  | Whether or not the worker has committed the initial snapshot for a source.
+`messages_received`   | [`bigint`]   | How many raw messages the worker has received from upstream.
+`updates`             | [`bigint`]   | How many updates (inserts + deletes) the worker has written but not yet committed to the storage layer.
+`updates_committed`   | [`bigint`]   | How many updates (inserts + deletes) the worker has committed into our storage layer.
+`bytes_received`      | [`bigint`]   | How many bytes worth of messages the worker has received from upstream. The way the bytes are counted is source-specific.
```
```suggestion
`messages_received`   | [`bigint`]   | The number of raw messages the worker has received from upstream.
`updates`             | [`bigint`]   | The number of updates (inserts + deletes) the worker has written but not yet committed to the storage layer.
`updates_committed`   | [`bigint`]   | The number of updates (inserts + deletes) the worker has committed into the storage layer.
`bytes_received`      | [`bigint`]   | The number of bytes worth of messages the worker has received from upstream. The way the bytes are counted is source-specific.
```

morsapaes commented on doc/user/content/sql/system-catalog/mz_internal.md:
```diff
@@ -471,6 +471,37 @@ Field              | Type       | Meaning
+
+These values all require some amount of interpretation:
+
+- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like postgres,
+  produce deletion messages that are counted the same as insertions.
+- The `updates*` metrics can be permanently lower or higher than `messages_received`, because:
+  - The storage layer can choose to consolidate messages within a certain time interval.
+  - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
+- The `messages_*` and `bytes_received` metrics can optionally be interpreted as rates that can loosely show
+  the changing speed of a source over time.
+- In general, the values should be summed across `worker_id`'s, and some metrics may only exist for some workers.
+  - Additionally, sources with multiple outputs (like postgres shards) will need `messages_committed`
+  and `snapshot_committed` summed/AND-ed across the top-level source and the connected tables.
```
```suggestion
- `messages_received` does not distinguish between inserts and deletes. Some upstream sources, like PostgreSQL,
  produce deletion messages that are counted the same as insertions.
- The `updates*` metrics can be permanently lower or higher than `messages_received`, because:
  - The storage layer can choose to consolidate messages within a certain time interval.
  - Some envelopes (like `UPSERT`) can produce 2 messages (an insert and a delete) from a single upstream message.
- The `messages_*` and `bytes_received` metrics can optionally be interpreted as rates that can loosely show
  the changing speed of a source over time.
- In general, the values should be summed across `worker_id`s, and some metrics may only exist for some workers.
  - Additionally, sources with multiple outputs (like PostgreSQL) will need `messages_committed`
  and `snapshot_committed` summed/AND-ed across the top-level source and its subsources.
```

morsapaes replied:
It's reasonable to keep these details here, for lack of a better place (at least for the time being).

morsapaes commented on src/adapter/src/catalog/builtin.rs:
```diff
@@ -1790,6 +1790,22 @@ pub static MZ_STORAGE_HOST_METRICS: Lazy<BuiltinSource> = Lazy::new(|| BuiltinSo
     is_retained_metrics_relation: true,
 });
 
+// This will be replaced with per-replica tables with source/sink multiplexing is supported.
```
```suggestion
// This will be replaced with per-replica tables once source/sink multiplexing on a single cluster is supported.
```

morsapaes reviewed (changes requested):
Left some documentation suggestions (mostly for consistency). Thanks for taking care of that in the PR, @guswynn!

guswynn commented:
@morsapaes @benesch updated with all your suggestions!

benesch reviewed (commented):
The new names and fields look 👌🏽 to me.

Could you split the docs changes out into a separate PR, though? There are some structural and phrasing improvements I'd like to suggest but we don't need to block this PR on it. We're also trying not to merge docs PRs before the release ships, so having the docs changes in a separate PR will allow us to keep the docs from going live until the release is out.

guswynn commented:
Will split the docs out!

aljoscha commented on src/storage-client/src/controller/statistics.rs:
```diff
@@ -0,0 +1,105 @@
+
+    let (shutdown_tx, mut shutdown_rx) = oneshot::channel::<()>();
+
+    mz_ore::task::spawn(|| "statistics_sender", async move {
```
nit: `"statistics_scraper"`?

aljoscha commented on src/storage/src/render/persist_sink.rs:
```diff
@@ -93,6 +93,13 @@ where
         current_upper.borrow_mut().clear();
     }
 
+    // All sub-sources increment the same stats object.
```
This commend is incorrect now, right? Because each sub source has their own stats object.

aljoscha commented on src/storage/src/source/statistics.rs:
```diff
@@ -0,0 +1,91 @@
+    // The boolean its an initialization flag, to ensure we
+    // don't report a false `snapshot_committed` value before
+    // the `persist_sink` reports the current shard upper.
+    stats: Rc<RefCell<(bool, SourceStatisticsUpdate)>>,
```
Still not sure I like that lone `bool` here. Maybe this could be a `Option<SourceStatisticsUpdate>`? But I also don't want to block merging this.

aljoscha reviewed (approved):
still liking this! I left some nits, but you might want to resolve those as follow-ups, or not at all. At least one of them is about an outdated comment, though.
